drivers/at_cmd/at_cmd_interface.c

```diff
@@ -0,0 +1,234 @@
+	u32_t         data_len;
+	char          data[RESPONSE_MAX_LEN];
+};
+
+static enum at_cmd_rc get_return_code(char *buf)
+{
+	enum at_cmd_rc return_code = AT_CMD_NO_RETURN_CODE;
+	char *tmpstr = NULL;
+
+	do {
+		tmpstr = strstr(buf, AT_CMD_OK_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_OK;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_ERROR_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_ERROR;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_CMS_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_CMS;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_CME_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_CME;
+			break;
+		}
+	} while (0);
+
+	if ((return_code == AT_CMD_CMS) || (return_code == AT_CMD_CME)) {
+		char tmpstr[5];
+
+		memcpy(tmpstr, &buf[strlen(AT_CMD_CMS_STR)], 3);
+		last_error = atoi(tmpstr);
+	} else {
+		last_error = 0;
+	}
+
+	if (tmpstr) {
+		buf[tmpstr - buf] = '\0';
+	}
+
+	return return_code;
+}
+
+static void callback_worker(struct k_work *item)
+{
+	struct callback_work_item *data =
+		CONTAINER_OF(item, struct callback_work_item, work);
+
+	received_data_handler(data->data, data->data_len);
+
+	k_free(data);
+}
+
+
+static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	int bytes_read;
+	char buf[RESPONSE_MAX_LEN];
+	enum at_cmd_rc return_code;
+	bool write_with_response = false;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	LOG_DBG("Socket thread started");
+
+	for (;;) {
+		bytes_read = recv(common_socket_fd, buf, sizeof(buf), 0);
+
+		if (bytes_read > RESPONSE_MAX_LEN) {
+			LOG_ERR("Please increase AT message reception buffer size");
+
+			/* Drop message as we can not handle the size */
+			continue;
+		}
+
+		return_code = get_return_code(buf);
+
+		if (return_code != AT_CMD_NO_RETURN_CODE) {
+			if (response_buf_len > 0) {
+				if (response_buf_len >= strlen(buf)) {
+					strncpy(response_buf, buf, bytes_read);
+					response_buf_len = 0;
+					write_with_response = true;
+				} else {
+					response_buf = NULL;
+				}
+			}
+
+			k_msgq_put(&return_code_msq, &return_code, K_FOREVER);
+		} else {
+			write_with_response = false;
+		}
+
+		if (strlen(buf) == 0) {
+			continue;
+		}
+
+		if (received_data_handler) {
+			if (!write_with_response) {
+				struct callback_work_item *item =
+				k_malloc(sizeof(struct callback_work_item));
+
+				if (!item) {
+					LOG_DBG("Failed to allocate work item");
+					continue;
+				}
+
+				k_work_init(&item->work, callback_worker);
+
+				item->data_len = strlen(buf);
+				memcpy(item->data, buf, item->data_len);
+
+				k_work_submit(&item->work);
+			}
+		}  else {
+			LOG_DBG("No handler registered for received data");
+		}
+	}
+}
+
+enum at_cmd_rc at_cmd_write(const char *const cmd)
+{
+	int bytes_sent;
+	enum at_cmd_rc return_code;
+
+	k_sem_take(&cmd_pending, K_FOREVER);
+
+	LOG_DBG("Sending command %s", cmd);
+	bytes_sent = send(common_socket_fd, cmd, strlen(cmd), 0);
+
+	k_msgq_get(&return_code_msq, &return_code, K_FOREVER);
```

Comment:
Why do we block here if we expect the response to be provided through callback function? I'd expect to block within `at_cmd_write_with_response`, synchronously waiting for response, not in `at_cmd_write`.

Reply:
Every AT command will always return a code OK, ERROR, +CMS: nnn, +CME: nnn", immediately  to let the caller know how the execution of the AT command went. I want to return this code back to the correct caller. At this level I don't have any means to figuring out which response is connected to which request. So if I have multiple threads sending AT commands at the same time I can loose track of which return code should be sent back to which caller.

So I do this to make sure that only one thread can send an AT command and get the return code at the time.

Reply:
I agree that we need to protect against multithreaded access somehow. The problem I see though is that simple blocking on a socket until response is received does not protect us at all.

Imagine we have 2 threads communicating with this module, and both wants to send some command. The flow could look like this in that case:
* Thread A sets the reponse handler with `at_cmd_set_handler` and calls `at_cmd_write`. This call will block waiting for response, kernel will reschedule threads.
* Thread B got CPU time while Thread A is still wating for response. It sets it's own response handler with `at_cmd_set_handler` and calls `at_cmd_write`. This call will block the thread.
* Now, the response desitined to Thread A arrived. `at_cmd_interface` thread will process it, and will call the registered data handler, which in this case belongs to Thread B, so we have a mismatch. Also, depending on threads priority, we can have either Thread A or Thread B awaken, it all depends on threads configuration.
* When the second response arrives, it will be again returned through the callback registered by Thread B. The remainig thread will be awaken.

My point is, that in order to allow a proper behavior in multithreaded environment (and I suppose that's the main purpose of this module), we need either a more advanced queueing mechanism, or in simplest approach a mutex protection of shared resources (like socket or response handler pointer).

And on more observation, from nRF Connect - modem seems to send periodically, let's say notifications, that are not responses - `%CESQ: 255,0` in my case. In current, these can be forwarded only to a single user of this module. I'm not sure whether it is important or not, but I guess it could be, as nRF Connect seems to process this data.

Reply:
I can copy the callback pointer while the code is in the callback_pending critical section, and submit that as part of work item. The user will  not be able to change callback handler while the driver is in this state. This will make sure that the intended callback is in fact called with the correct response.

@metc Proposed to drop the whole dynamic callback handler support and instead have a compile-time static handler. Maybe that is not such a bad idea after all, when we start to look into this multi-thread scenarios. The goal post is anyway to create a service on top of this driver and the parser-decoder in PR606 which will let multiple clients subscribe to data in the AT command notifications and responses, but will make the AT interface completely transparent.

Reply:
In the scenario described by @rlubos, I do not think we are allowed to write a new AT command before the response of the previous one has been received. Some AT command will return a response immediately (simple AT commands like reading manufacturer) but most commands will not respond immediately. In that case, Thread B should wait until Thread A got a response from that socket. And during that time, it should not be possible to change the handler.

@evenl I think registering the handler function at initialization or compilation time is good enough. I do not have a use case where we need to change the handler runtime.

Reply:
But we already have multiple users of AT sockets: `lte_lc`, `at_host` and `modem_info`, and from what I understood, the goal was to allow them all to use a single socket through this interface. If we limit the usage of this interface to a single library, what's the point of it?

If we just want to provide a convenience library that would simplify the AT communication and do not want to burden with multithreaded access, we should not use a single socket and global context data, but let the user of this library to create a context for the library, and provide each API call with this context. This approach would be thread-safe, would not limit the number of users of this library, but as a tradeoff, does not save us any resources (sockets).

Reply:
I think my scheme is properly protected in multi-thread scenarios now. I have also ported lte_lc and modem_info to also use the at_interface driver.